### PR TITLE
unit tests for attrs

### PR DESCRIPTION
**Pull request by davidangb**
See also broadinstitute/firecloud-orchestration#274
- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Make sure Swagger is updated if API changes
- [x] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [x] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [x] **Submitter**: JIRA ticket checks:
  - Acceptance criteria exists and is met
  - Note any changes to implementation from the description
  - Add notes on what you've tested
- [x] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [x] **Submitter**: Database checks:
  - If PR includes new or changed db queries, include the explain plans in the description
  - Make sure liquibase is updated if appropriate
  - If doing a migration, take a backup of the
    [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
    and
    [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
    DBs in Google Cloud Console
- [x] **Submitter**: Update FISMA documentation if changes to:
  - Authentication
  - Authorization
  - Encryption
  - Audit trails
- [x] **Submitter**: If you're adding new libraries, sign us up to security updates for them
- [x] Tell your tech lead (TL) that the PR exists if they want to look at it
- [x] Anoint a lead reviewer (LR). **Assign PR to LR**
- Review cycle:
  - LR reviews
  - Rest of team may comment on PR at will
  - **LR assigns to submitter** for feedback fixes
  - Submitter rebases to develop again if necessary
  - Submitter makes further commits. DO NOT SQUASH
  - Submitter updates documentation as needed
  - Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via hipchat and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: src/test/scala/org/broadinstitute/dsde/rawls/model/AttributeSpec.scala [hunk 1]
```diff
@@ -0,0 +1,489 @@
...
+        val expected = AttributeValueList(Seq(
+          AttributeBoolean(true),
+          AttributeBoolean(false),
+          AttributeBoolean(true)
+        ))
+        assertResult(expected) { attr }
+      }
+
+      // TODO: either re-enable these tests once target code throws a better exception,
+      // TODO: or change the tests to intercept ClassCastException
```
Comment (helgridly) [inline src/test/scala/org/broadinstitute/dsde/rawls/model/AttributeSpec.scala]: Good catch. I think this should be a DeserializationException.

File: src/test/scala/org/broadinstitute/dsde/rawls/model/AttributeSpec.scala [hunk 2]
```diff
@@ -0,0 +1,489 @@
...
+            |   ]
+            | }}
+          """.stripMargin
+        intercept[RawlsException] {
+          testData.parseJson.convertTo[AttributeMap]
+        }
+      }
+      // TODO: end ignored tests
+
+      // TODO: should this next case work as well as it does?
```
Comment (helgridly) [inline src/test/scala/org/broadinstitute/dsde/rawls/model/AttributeSpec.scala]: this is legal and has been forever, so yes

though why i have no idea

File: src/test/scala/org/broadinstitute/dsde/rawls/model/AttributeSpec.scala [hunk 3]
```diff
@@ -0,0 +1,489 @@
...
+        val obj = testData.parseJson.convertTo[AttributeMap]
+        val attr = obj.getOrElse(AttributeName.withDefaultNS("testKey"), fail("attr doesn't exist"))
+        val expected = AttributeValueList(Seq(
+          AttributeString("hello world"),
+          AttributeNumber(123),
+          AttributeBoolean(false)
+        ))
+        assertResult(expected) { attr }
+      }
+      // TODO: should this next case throw DeserializationException or something else?
```
Comment (helgridly) [inline src/test/scala/org/broadinstitute/dsde/rawls/model/AttributeSpec.scala]: I think DeserializationException is correct, but it should bubble up to the user as a 400

**Comment by helgridly** [comment]
A couple of comments re your todos.

Could you throw in a test at the *Service level to check that a DeserializationException when e.g. updating an entity's attributes correctly turns into a 400?

**Comment by davidangb** [comment]
@helgridly back to you, a few updated tests - the new tests will fail!!!

**Comment by helgridly** [comment]
Let's do it! Thanks for your help here.

**Pull request merged**